eeps/eep-0064.md

```diff
@@ -0,0 +1,187 @@
+    Author: Kiko Fernandez-Reyes <kiko(at)erlang(dot)org>
+    Status: Draft
+    Type: Standards Track
+    Created: 07-Jun-2023
+    Erlang-Version: OTP-27
+    Post-History:
+****
+EEP XXX: Triple-quoted binary strings
+----
+
+Abstract
+========
+
+This EEP proposes the introduction of triple-quoted binary-strings syntax and
+runtime semantics. The main benefit is to allow multi-line binary strings in an
+easy way, similar to other languages, e.g., Elixir.
+
+Rationale
+=========
+
+One limitation of Erlang today (June 2023) is the ability to write multi-line
+string binaries. That said, the main reason to consider this EEP is
+[EEP-59][] where documentation attributes can
+benefit from multi-line string binaries: the documentation generates Docs chunks
+format -- documentation in its binary format-- which saves space and makes the
+documentation available to the shell.
+
+Triple-Quoted Binary-String Design Decisions
+---------------------------------------------
+
+### Runtime semantics
+
+Triple-quoted strings should only produce binary-strings. This makes easy to
```

Comment:
I worry this may be potentially confusing. Why is `"foo"` a list but `"""foo"""` a binary? What if I want to write a long text but as a charlist instead? Here is [a potential example](https://github.com/elixir-lang/elixir/blob/0872ee5444f257744866a4d4c237057f2d97d601/lib/elixir/src/elixir_parser.yrl#L1237-L1249). So, even though I would prefer binaries, I believe a more consistent option is to return charlists.

Then, for binaries, there are a few options:

1. Write `<<"""...""">>`

2. Build on EEP 63 to introduce `u"""..."""`

3. However, for documentation in particular, the compiler can convert `-doc """...""".` into binaries, so the distinction may not be terribly important

Reply:
Completely agreeing with José here, I think triple quotes strings should produce strings in erlang unless I closed in a binary context, this would be more consistent with the existing Erlang syntax.

Reply:
I agree with José as well, I think it is confusing to use `""""` as a binary. Another option for unicode binaries could be to use backticks:

    io:format(`backtick unicode binary`)
    -doc ```
           backtick 
           multiline
           unicode
           binary
         ```.
Not the most convenient key on some keyboards, but Javascript seems to get away with using it so maybe we can aswell.

Reply:
I like using backticks to mean binary. That nicely solves how to write binary literal strings on one line as well. I think we have a winner!

Reply:
Backticks is interesting because it also exists in Prolog. Erlang's syntax comes from Prolog as we all know (which also has single quote for atoms). In SWI Prolog it seems to be configurable how double-quoted and back-quoted strings work: https://www.swi-prolog.org/pldoc/man?section=string and in [GNU Prolog](http://gprolog.org/manual/gprolog.html#set-prolog-flag%2F2) similarily. 

But, for simple binary strings, isn't a prefix like `b"binary string"` simpler and more aligned with Python, C, Rust, etc. for different kinds of strings?

I think `b""` is enough for UTF-8 encoded binaries as long as the source code is UTF-8 encoded, which it normally is. (If we then add triple quotes for charlist strings, then `b"""..."""` seems the natural choice for the binary version.)

(Btw [C11](https://en.cppreference.com/w/c/language/string_literal) uses `u""` for UTF-16 strings, `U""` for UTF-32 strings and `u8""` for UTF-8, so `u""` for UTF-8 would be confusing in this context.)

Reply:
There are currently three proposals for revamping strings in Erlang:

* Adding string interpolation
* Adding a shortcut notation for utf-8 encoded binaries
* Adding support for heredocs/triple-quoted strings

The tricky part is: we cannot consider each of them in isolation. `"charlists"` cannot receive string interpolation today because that would be backwards incompatible. If we introduce `` `utf-8 encoded binaries` ``, if they are to receive interpolation, the interpolation syntax must be reserved now. At the same time, I feel it would be inconsistent if utf-8 encoded binaries (backticks) get interpolation and strings (double-quoted) do not, so I feel like choosing backticks would close the door on string interpolations. Which is a fine decision to make as long as we are intentional about it.

The argument for `u"..."` style is that it provides a general syntax. For example, regexes are tricky to write in Erlang because you need to consistently double escape. A `r"..."` syntax would solve this problem (and we could precompile regexes too, instead of calling `re:compile/1`). We could even make it user-extensible in the future, similar to how it works in JavaScript, allowing a developer to have their own `merl"..."` or `xml"..."` strings.

I am not necessarily arguing in favor of all of these extensions but my point is that `xyz"..."` provides room for growth. Even if we say: we don't want to have string interpolation now, will that still be true in 5 years or 10 years? The upcoming Java 21 is getting string interpolation.

Finally, choosing backtick will also depend on the documentation format. If we choose Markdown, the backticks in Markdown are used for code snippets. Which means that, if I want to write "string:to_upper(`hello`)" in the docs, I will have to surround it by double backticks, like this: ``` ``string:to_upper(`hello`).`` ``` and I am afraid this will frequently trip people up.

---

Overall, my vote would be the following:

1. Introduce triple-quoted strings as in this proposal but have them return lists (this means we don't care about binaries for now)

2. Because `"foo"` does not have interpolation, it is expected `"""foo"""` does not have interpolation either (this means we don't care about interpolation for now)

3. Allow `-doc` to receive either binary strings or list strings and document they are normalized to binary strings

This allows us to side-step the other proposals for now. Then we should discuss if-when-how to add interpolation and if-when-how to add a syntax shortcut for binaries.

Reply:
We can introduce interpolation in all strings, by using an improbable escape sequence: `\(`...`)` in any string.
Then would choosing backticks for binary strings not close the door for interpolation.

Reply:
As a side note, that improbable string sequence specifically (`\(...)`) would absolutely risk clashing with a bunch of regexes though, as a way to match on a parens without creating a group.

Reply:
Yes, but a regexp string is a different animal where I guess you do not want any escape sequences, and therefore cannot use interpolation. Possibly chosen with a prefix, e.g `r"foo\(\.\)+"`.

Reply:
I like that Erlang becomes more Perl-like. :-)

Reply:
> Finally, choosing backtick will also depend on the documentation format. If we choose Markdown, the backticks in Markdown are used for code snippets. Which means that, if I want to write ``string:to_upper(`hello`)``

This will be also problematic with heredocs and fenced code blocks:

<img width="178" alt="image" src="https://github.com/erlang/eep/assets/76071/69f719c4-96a1-4f2a-8fd9-cdc44c1a0570">

(I'm pasting an imagine because I don't know how to write it in Markdown so it renders correctly. :))

I'm personally a fan of code blocks done with 4 spaces indent but fenced blocks are sometimes useful too.

Reply:
@wojtekmach Markdown allows any number of backticks, so if you need 3 backticks, just write it within 4 backticks. Perhaps Erlang could allow that trick too:

`````
-doc ````
Some function.

Example:

```
> 1 + 2
3
```
````
`````

(Note that the above example was enclosed in 5 backticks.)

Reply:
I like the backtick idea, but I'll prefer it to be used like in Javascript, where it permits to do interpolation, for example:
```
Bar = `bar`,
<<"foo bar">> = `foo ${Bar}`.
```
or 
````
-define(BAR, `bar`).

-doc ```
foo ${?BAR}
```
````
Or (why not) single backtick as an alternative to triple quotes using the same idea, the last backtick defines the indentation
```
-doc `
backtick 
multiline
unicode
binary
`

% Single line also a valid syntax
-doc `backtick unicode binary`
```

BTW, triple quotes give more control of the encoding (I don't see any problem writing the below)

```
<<"""
foo bar
"""/utf8>>
```